Source/FileSystem/ModuleResources.cpp

```diff
@@ -563,70 +570,76 @@ void ModuleResources::MonitorResources()
 	{
 		ModuleFileSystem* fileSystem = App->GetModule<ModuleFileSystem>();
 		CreateAssetAndLibFolders();
+
 		std::vector<std::shared_ptr<EditorResourceInterface>> toRemove;
 		std::vector<std::shared_ptr<Resource>> toImport;
 		std::vector<std::shared_ptr<Resource>> toCreateLib;
 		std::vector<std::shared_ptr<Resource>> toCreateMeta;
-		for (auto resourceit = resources.begin(); resourceit != resources.end();)
+
+		// create a copy of the current state of the resources cache to avoid concurrency problems
+		std::map<UID, std::weak_ptr<Resource>> copyOfResources = resources;
+
+		for (const auto& [uid, weakResource] : copyOfResources)
 		{
-			const std::shared_ptr<Resource>& resource = resourceit->second.lock();
-			if (resource)
+			std::shared_ptr<Resource> resource = weakResource.lock();
+			if (resource == nullptr)
 			{
-				if (resource->GetType() != ResourceType::Mesh && !fileSystem->Exists(resource->GetAssetsPath().c_str()))
+				continue;
+			}
+
+			if (resource->GetType() != ResourceType::Mesh && !fileSystem->Exists(resource->GetAssetsPath().c_str()))
+			{
+				toRemove.push_back(std::dynamic_pointer_cast<EditorResourceInterface>(resource));
+			}
+			else
+			{
+				std::string libraryPathWithExtension = fileSystem->GetPathWithExtension(resource->GetLibraryPath());
+
+				if (libraryPathWithExtension.empty() /*file with that name was not found*/ ||
+					!fileSystem->Exists(libraryPathWithExtension.c_str()))
 				{
-					toRemove.push_back(std::dynamic_pointer_cast<EditorResourceInterface>(resource));
+					toCreateLib.push_back(resource);
 				}
-				else
+				if (!fileSystem->Exists((resource->GetAssetsPath() + META_EXTENSION).c_str()))
 				{
-					std::string libraryPathWithExtension = fileSystem->GetPathWithExtension(resource->GetLibraryPath());
-
-					if (libraryPathWithExtension.empty() /*file with that name was not found*/ ||
-						!fileSystem->Exists(libraryPathWithExtension.c_str()))
-					{
-						toCreateLib.push_back(resource);
-					}
-					if (!fileSystem->Exists((resource->GetAssetsPath() + META_EXTENSION).c_str()))
-					{
-						toCreateMeta.push_back(resource);
-					}
-					// these type's assets are binary files changed in runtime
-					else if (resource->GetType() != ResourceType::Mesh && resource->GetType() != ResourceType::Material)
+					toCreateMeta.push_back(resource);
+				}
+				// these type's assets are binary files changed in runtime
+				else if (resource->GetType() != ResourceType::Mesh && resource->GetType() != ResourceType::Material)
+				{
+					long long assetTime = fileSystem->GetModificationDate(resource->GetAssetsPath().c_str());
```

Comment:
```suggestion
					long long assetTimestamp = fileSystem->GetModificationDate(resource->GetAssetsPath().c_str());
```
not sure if needed, but I think this name is more accurate if we are looking for the exact date and time of the file modified